incubator/hnc/pkg/validators/hncconfig.go

```diff
@@ -0,0 +1,182 @@
+			return allow("")
+		}
+	}
+
+	inst := &api.HNCConfiguration{}
+	if err := c.decoder.Decode(req, inst); err != nil {
+		c.Log.Error(err, "Couldn't decode request")
+		return deny(metav1.StatusReasonBadRequest, err.Error())
+	}
+
+	resp := c.handle(ctx, inst)
+	c.Log.Info("Handled", "allowed", resp.Allowed, "code", resp.Result.Code, "reason", resp.Result.Reason, "message", resp.Result.Message)
+	return resp
+}
+
+// handle implements the validation logic of this validator for Create and Update operations,
+// allowing it to be more easily unit tested (ie without constructing a full admission.Request).
+func (c *HNCConfig) handle(ctx context.Context, inst *api.HNCConfiguration) admission.Response {
+	if inst.GetName() != "config" {
+		return deny(metav1.StatusReasonInvalid, fmt.Sprintf("Wrong singleton name: %s; the name should be 'config'", inst.GetName()))
+	}
+
+	roleExist := false
+	roleBindingExist := false
+	ts := gvkSet{}
+	for _, t := range inst.Spec.Types {
+		if rp := c.isTypeConfigured(t, ts); !rp.Allowed {
+			return rp
+		}
+
+		if c.isRBAC(t, "Role") {
+			if rp := c.validateRBAC(t.Mode, "Role"); !rp.Allowed {
+				return rp
+			}
+			roleExist = true
+		} else if c.isRBAC(t, "RoleBinding") {
+			if rp := c.validateRBAC(t.Mode, "RoleBinding"); !rp.Allowed {
+				return rp
+			}
+			roleBindingExist = true
+		} else {
+			if rp := c.validateType(ctx, t, ts); !rp.Allowed {
+				return rp
+			}
+		}
+	}
+	if !roleExist {
+		return deny(metav1.StatusReasonInvalid, "Configuration for Role is missing")
+	}
+	if !roleBindingExist {
+		return deny(metav1.StatusReasonInvalid, "Configuration for RoleBinding is missing")
+	}
+	return allow("")
+}
+
+// Validate if the configuration of a type already exists. Each type should only have one configuration.
+func (c *HNCConfig) isTypeConfigured(t api.TypeSynchronizationSpec, ts gvkSet) admission.Response {
+	gvk := schema.FromAPIVersionAndKind(t.APIVersion, t.Kind)
+	if exists := ts[gvk]; exists {
+		return deny(metav1.StatusReasonInvalid, fmt.Sprintf("Duplicate configurations for %s", gvk))
+	}
+	ts[gvk] = true
+	return allow("")
+}
+
+func (c *HNCConfig) isRBAC(t api.TypeSynchronizationSpec, kind string) bool {
+	return t.APIVersion == "rbac.authorization.k8s.io/v1" && t.Kind == kind
+}
+
+// The mode of Role and RoleBinding should be either unset or set to the propagate mode.
+func (c *HNCConfig) validateRBAC(mode api.SynchronizationMode, kind string) admission.Response {
+	if mode == api.Propagate || mode == "" {
+		return allow("")
+	}
+	return deny(metav1.StatusReasonInvalid, fmt.Sprintf("Invalid mode of %s; current mode: %s; expected mode %s", kind, mode, api.Propagate))
+}
+
+// validateType validates a non-RBAC type.
+func (c *HNCConfig) validateType(ctx context.Context, t api.TypeSynchronizationSpec, ts gvkSet) admission.Response {
+	gvk := schema.FromAPIVersionAndKind(t.APIVersion, t.Kind)
+
+	// Validate if the GVK exists in the apiserver.
+	exists, err := c.validator.Exists(ctx, gvk)
```

Comment:
Nit: is `err` always non-nil if `exists` is false? If not, why not:

```
if err := c.validator.Exists(ctx, gvk); err != nil {
  return deny(...)
}
```

Reply:
It is possible that err returned by Exists is `errors.IsNotFound`. In this case exists is true. It means the GVK exists in the cluster but the object is not found. It is likely possible that the object does not exist, since we are querying a random object of the GVK. If GVK does not exist in the cluster, the error is not nil but also not `errors.IsNotFound`

See line: 168.

Reply:
That seems like a bit of an odd implementation detail to leak out of this function. The caller of `c.validator.Exists` isn't asking if an arbitrarily named object exists, so returning an error that says "not found" doesn't really seem relevant? I'd rather have something inside Exists like:

```
if err == errors.IsNotFound {
  return nil // the type exists
}
return err // if err is nil, that means the object was found, which means the type exists
```

Alternatively, it could _just_ return the boolean, since the text of the error message itself is unlikely to be interesting. In which case the logic will be:

```
if err == nil || err == errors.IsNotFound {
  return true // the type exists
}
return false
```

Reply:
Thanks for the suggestion! Only returning an error makes sense. I will still output the error message just in case users are interested and in case there are some transient errors occur when querying the apiserver. See #525